### PR TITLE
add doxygen coverage

### PR DESCRIPTION
**Pull request by junghans**

### CODE REVIEW

**Comment by junghans** [comment]
@sslattery @dalg24 we broke doxygen's xml output:
```
error: error while parsing xml file xml/classCabana_1_1AoSoA.xml : not well-formed (invalid token): line 25, column 33
```

**Comment by junghans** [comment]
apparently there was a bug in doxygen, so this PR won't be able to merged soon. Alternatively we need to build our own doxygen, which might not be worth it.

**Comment by junghans** [comment]
Or we could patch the xml files using `sed`.

**Comment by sslattery** [comment]
@junghans do you still want a review here if we are unable to merge due to the bug?

**Comment by junghans** [comment]
@sslattery, I am working on workaround...

**Comment by junghans** [comment]
@dalg24, 20.4% doxygen coverage isn't great, but at least we have the numbers now!

**Comment by junghans** [comment]
@dalg24 please review

**Comment by sslattery** [comment]
@junghans what's up with the coverage changes?

**Comment by sslattery** [comment]
Is it that doxygen is now included in the coverage?

File: .travis.yml [hunk 1]
```diff
@@ -53,6 +53,10 @@ before_script:
       popd;
     fi
   - for i in ${BACKENDS}; do CMAKE_OPTS+=( -DCabana_ENABLE_${i}=ON ); done
+  - if [[ ${COVERAGE} ]]; then
+       pip install --user coverxygen &&
+       mkdir -p $HOME/.local/bin && wget -O $HOME/.local/bin/codecov https://codecov.io/bash && chmod +x $HOME/.local/bin/codecov;
```
Comment (sslattery) [inline .travis.yml]: OK so here it looks like you are adding doxygen to the coverage - correct?

File: .travis.yml [hunk 2]
```diff
@@ -82,12 +87,17 @@ after_success:
  - ccache -s
  - test "${TRAVIS_BUILD_STAGE_NAME}" = "Test" || travis_terminate 0
  - if [[ ${COVERAGE} ]]; then
-     pushd build && wget -O codecov https://codecov.io/bash && chmod +x codecov &&
+     pushd build &&
      if [[ ${CC} = clang ]]; then
-       ./codecov -x "llvm-cov gcov" -F "${CC}";
+       codecov -x "llvm-cov gcov" -F "${CC}";
```
Comment (sslattery) [inline .travis.yml]: What is different about coverage with LLVM?

File: .travis.yml [hunk 3]
```diff
@@ -82,12 +87,17 @@ after_success:
...
      else
-       ./codecov -F "${CC}";
+       codecov -F "${CC}";
      fi &&
+     if [[ $(doxygen --version) != 1.8.16 ]]; then
+       echo "Patching Doxygen files (see doxygen/doxygen#6892)" &&
+       sed -e '/enable_if</s/&/\&amp;/g' -e '/enable_if/s/<\([^a-z/]\)/\&lt;\1/g' -e '/enable_if/s/ >/ \&gt;/g' -i xml/*.xml;
```
Comment (sslattery) [inline .travis.yml]: Here it looks like you are removing the offending code in the XML files? Is 1.8.16 the latest doxygen version that has the bug fix?

File: CMakeLists.txt [hunk 4]
```diff
@@ -62,6 +62,7 @@ endif()
 
 find_package(Doxygen)
 if(Doxygen_FOUND)
+  set(DOXYGEN_GENERATE_XML YES)
```
Comment (sslattery) [inline CMakeLists.txt]: So ultimately coverage with Doxygen requires coverage to assess the doxygen-generated XML files?

Comment (junghans) [reply CMakeLists.txt]: Yes, it needs the xml outputs!

File: .travis.yml (hunk 3 above)
Comment (junghans) [reply .travis.yml]: Yes, the fix is part of the next release, 1.8.16.

File: .travis.yml (hunk 2 above)
Comment (junghans) [reply .travis.yml]: For `clang` you need to use `llvm-cov gcov` instead of `gcov`.

**Comment by junghans** [comment]
Yes, see the percentage for `#doxygen` Flag above.

File: .travis.yml (hunk 1 above)
Comment (junghans) [reply .travis.yml]: `coverxygen` is the new tool. I download codecov here as it seems more like a `before_script:` step.

**Comment by junghans** [comment]
And coverage check is failing now as doxygen pulls the overall coverage down.

**Comment by dalg24** [comment]
Please post a link to the Doxygen coverage report.  I browsed a bit on codecov and could not make any sense of what I saw.

**Comment by junghans** [comment]
> Please post a link to the Doxygen coverage report. I browsed a bit on codecov and could not make any sense of what I saw.

Yes, it is confusing to find, you need to go to the file view and then click on “doxygen” at the very top. Unfortunately  there is no summary view per tags, only per file.

File: .travis.yml (hunk 3 above)
Comment (sslattery) [reply .travis.yml]: should we use `<` instead of `!=` for the version check assuming that future versions will also not have this bug?

**Comment by junghans** [comment]
@sslattery version comparison is hard:
```
$ [[ 1.8.16 < 1.8.2 ]] && echo error
error
```
I did `[[ $(doxygen --version) != 1.8.16 ]]` as Doxygen release are slow and we can delete the whole block anyhow once doxygen is up to date.

**Comment by junghans** [comment]
@sslattery #112

**Pull request merged**